src/procedures/fulltext_create_index.c

```diff
@@ -17,13 +18,113 @@
+    Graph *g,
+    int label_id,
+    char **fields,
+    uint *fields_ids,
+    uint fields_count)
+{
+    Node node;
+    NodeID node_id;
+    GraphEntity *entity;
+    EntityProperty *prop;
+    GxB_MatrixTupleIter *it;
+    const GrB_Matrix label_matrix = Graph_GetLabel(g, label_id);
+    GxB_MatrixTupleIter_new(&it, label_matrix);
+
+    while(true) {
+        bool depleted = false;
+        GxB_MatrixTupleIter_next(it, NULL, &node_id, &depleted);
+        if(depleted) break;
+        
+        double score = 0;
+        const char* lang = NULL;
+        Graph_GetNode(g, node_id, &node);
+        RSDoc *doc = RediSearch_CreateDocument(&node_id, sizeof(EntityID), score, lang);
+
+        for(uint i = 0; i < fields_count; i++) {
+            RedisModuleString* s;
+            SIValue *v = GraphEntity_GetProperty((GraphEntity*)&node, fields_ids[i]);
+            if(v == PROPERTY_NOTFOUND) continue;
+            if(v->type & SI_NUMERIC) continue;
+            RediSearch_DocumentAddFieldString(doc, fields[i], v->stringval, strlen(v->stringval), RSFLDTYPE_FULLTEXT);
+        }
+
+        RediSearch_SpecAddDocument(idx, doc);
+    }
+
+    GxB_MatrixTupleIter_free(it);
+}
+
+static int _getNodeAttribute(void* ctx, const char* fieldName, const void* id, char** strVal, double* doubleVal) {
+    Node n;
+    NodeID nId = *(NodeID*)id;
+    _fulltextIndexCtx *indexCtx = ctx;
+    Graph *g = indexCtx->g;
+    Schema *s = indexCtx->s;
+    
+    Graph_GetNode(g, nId, &n);
+    Attribute_ID attrId = Schema_GetAttributeID(s, fieldName);
+    SIValue *v = GraphEntity_GetProperty((GraphEntity*)&n, attrId);
+    int ret;
+    if(v == PROPERTY_NOTFOUND) {
+        ret = RSVALTYPE_NOTFOUND;
+    } else if(v->type & SI_STRING) {
+        *strVal = v->stringval;
+        ret = RSVALTYPE_STRING;
+    } else {
+        *doubleVal = v->doubleval;
+        ret = RSVALTYPE_DOUBLE;
+    }
+    return ret;
+}
+
 ProcedureResult fulltextCreateNodeIdxInvoke(ProcedureCtx *ctx, char **args) {
     if(array_len(args) < 2) return PROCEDURE_ERR;
     
+    // Create full-text index.
     char *label = args[0];
-    for(int i = 0; i < array_len(args); i++) {
-        // Create full-text index.
+    GraphContext *gc = GraphContext_GetFromTLS();
+    _fulltextIndexCtx *indexCtx = rm_malloc(sizeof(_fulltextIndexCtx));
+    Graph *g = gc->g;
+    Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
+    // Schema doesn't exists, TODO: report error.
+    if(!s) return PROCEDURE_ERR;
+
+    indexCtx->g = g;
+    indexCtx->s = s;
+    int label_id = s->id;
+    uint fields_count = array_len(args) - 1;
+    uint fields_ids[fields_count];
+    char **fields = args+1; // Skip index name.
+
+    // Validate that all specified fields exists.
+    for(int i = 0; i < fields_count; i++) {
+        char *field = fields[i];
+        if(Schema_GetAttributeID(s, field) == ATTRIBUTE_NOTFOUND) {
+            // TODO: report error.
+            return PROCEDURE_ERR;
+        }
+    }
```

Comment:
Loop will be outdated by `schema->attributes` removal.

Reply:
👍